packages/vite/src/node/server/moduleGraph.ts

```diff
@@ -215,4 +247,145 @@ export class ModuleGraph {
+          return
+        }
+        const module = await this.ensureEntryFromUrl(url)
+        ensureWatchedFile(this.watcher, module.file, this.config.root)
+        module.sourceEtag = value.sourceEtag
+        module.transformResult = value.transformResult
+      }
+    })
+    isDebug &&
+      debugModuleGraph(
+        timeFrom(start),
+        `${this.urlToModuleMap.size}/${
+          Object.keys(cache.files).length
+        } modules restored`
+      )
+  }
+
+  queueSaveCache(): void {
+    if (!this.saveCacheState.canQueue) return
+    if (this.saveCacheState.timeout) clearTimeout(this.saveCacheState.timeout)
+    this.saveCacheState.timeout = setTimeout(() => {
+      this.saveCacheState.timeout = undefined
+      this.saveCache()
+    }, 500)
+  }
+
+  async close(save: boolean = true): Promise<void> {
+    if (this.saveCacheState.timeout) clearTimeout(this.saveCacheState.timeout)
+    this.saveCacheState.canQueue = false
+    if (save) return this.saveCache()
+  }
+
+  private async saveCache(): Promise<void> {
+    if (!this.saveCacheState.promise) {
+      this.saveCacheState.promise = this.unsafelySaveCache().finally(() => {
+        this.saveCacheState.promise = undefined
+      })
+    }
+    return this.saveCacheState.promise
+  }
+
+  private async unsafelySaveCache(): Promise<void> {
+    const start = isDebug ? performance.now() : 0
+    const cacheLocation = this.getCacheLocation()
+    if (!cacheLocation) {
+      isDebug && debugModuleGraph('No cache location, saveCache skipped')
+      return
+    }
+    const cache: Cache = { configHash: this.getCacheHash(), files: {} }
+    this.urlToModuleMap.forEach((module, url) => {
+      if (!module.transformResult) return
+      if (
+        !module.sourceEtag ||
+        url.includes('node_modules') ||
+        /[?&]html-proxy\b/.test(url) || // Caching html proxy would require re-populate the htmlProxyMap
+        url.startsWith(FS_PREFIX) ||
+        url.startsWith(CLIENT_PUBLIC_PATH)
+      ) {
+        return
+      }
+      cache.files[url] = {
+        sourceEtag: module.sourceEtag,
+        transformResult: module.transformResult
+      }
+    })
+    await fs.promises.writeFile(cacheLocation, JSON.stringify(cache))
+    isDebug &&
+      debugModuleGraph(
+        timeFrom(start),
+        `${Object.keys(cache.files).length}/${
+          this.urlToModuleMap.size
+        } modules saved`
+      )
+  }
+
+  private getCacheHash(): string {
+    return getLockAndConfigHash(this.config.root, {
+      mode: this.config.mode,
+      root: this.config.root,
+      resolve: this.config.resolve,
+      esbuild: this.config.esbuild,
+      css: this.config.css,
+      define: this.config.define,
+      env: this.config.env,
+      configFileHash: this.config.configFileHash
+    })
+  }
+
+  private getCacheLocation(): string | undefined {
+    if (!this.config.cacheDir) return
+    return path.resolve(this.config.cacheDir, 'cache.json')
```

Comment:
Currently internal files in cacheDire are named with `_` prefix (`_metadata.json`, `_cert.pem`), maybe worth following ?
(On a personal touch I'd go with a more descriptive name, maybe `_moduleCache.json`)